### ISSUE TITLE
Identify hardware requirements for TEE tests

### ISSUE THREAD

**Issue by wainersm** [confidential-containers/community#4]
We aim to implement Kata Containers specifics tests for Confidential Containers for features which involve specialized hardware (i.e. TEE tests). One example of TEE specific feature is the confidential guest (property `confidential_guest=true` set on kata's *configuration.toml*) which utilizes hardware technologies such as *AMD SEV* to achieve memory and cpu-state encryption.  

Currently on Kata CI we use mostly Azure VMs that aren't capable for running those TEE specific tests. Bare-metal machines can easily be added on Kata's Jenkins but not all hardware vendors involved on this project can expose their machine in such as setup. For example, some may require to have the machine behind a firewall.
 
Therefore, this issue was created so that those hardware vendors can put their requirements and needs, and the community can discuss solutions to allow TEE tests can be ran on their machines.

Cc @hbrueckner @peterzcst @Jakob-Naucke migrating to this issue the discussion we started on https://github.com/confidential-containers/community/issues/1

**Comment by peterzcst** [comment]
Thanks @wainersm . We would provide a Intel TDX machine behind a firewall. By far ,we already made some progress. 
1. We already install a Jenkins master in an TDX machine  behind a firewall. 
2. We already implement proxy notifications for push/PR events from external github to internal Jenkins master to trigger CI/CD jobs. 
3. The CI jobs for kata containers master builds with TDX enabled guest kernel/qemu can successfully execute

More to coming to reach final goal like:
1. Enable CI jobs for PR builds 
2. Enable CI jobs for different branches and CC related repos
3. CI jobs for cloud-hypervisor with td-shim
4. Fix some of the broken k8s integration tests
5. integrate with non-TEE specific CC CI jobs
6. And of course we need define which tests should go to TEE machines 
7. maybe more...

**Comment by wainersm** [comment]
Hi @peterzcst , very interesting your setup.  

> Thanks @wainersm . We would provide a Intel TDX machine behind a firewall. By far ,we already made some progress.
> 
>     1. We already install a Jenkins master in an TDX machine  behind a firewall.

I'm just curious why the master needs to be in the TDX machine. Or maybe it is not required but you just made it to be more convenient.

> 
>     2. We already implement proxy notifications for push/PR events from external github to internal Jenkins master to trigger CI/CD jobs.

Got it. Does the proxy forward notifications both-ways (i.e. the PR will be notified with the build result)?

@Jakob-Naucke are you thinking in a similar setup for your s390x machine?

> 
>     3. The CI jobs for kata containers master builds with TDX enabled guest kernel/qemu can successfully execute
>

Sorry, I am still learning about the CC implementation on Kata. By `TDX enabled guest kernel/qemu` you mean  simply `confidential_guest = true` on Kata's `configuration.toml`?
 
> 
> More to coming to reach final goal like:
> 
>     1. Enable CI jobs for PR builds
> 
>     2. Enable CI jobs for different branches and CC related repos
> 
>     3. CI jobs for cloud-hypervisor with td-shim
> 
>     4. Fix some of the broken k8s integration tests
> 
>     5. integrate with non-TEE specific CC CI jobs

Recently we have [merged a non-TEE specific CC test](https://github.com/kata-containers/tests/pull/4508) and I'm working to get a [new job](https://github.com/kata-containers/ci/pull/446) running it. Maybe you can use those as references.

And the list of non-TEE tests is [here](https://github.com/confidential-containers/community/issues/3). We "only" need to implement them all :D

> 
>     6. And of course we need define which tests should go to TEE machines

I just created the issue #5  so that we can start discussing ideas of tests for TEE features.

> 
>     7. maybe more...

**Comment by Jakob-Naucke** [comment]
> > 2. We already implement proxy notifications for push/PR events from external github to internal Jenkins master to trigger CI/CD jobs.
> > ```
> 
> Got it. Does the proxy forward notifications both-ways (i.e. the PR will be notified with the build result)?
> 
> @Jakob-Naucke are you thinking in a similar setup for your s390x machine?

@wainersm no, I should think we'd just set up a slave instead which merely requires a tunnel from the master.

**Comment by peterzcst** [comment]
> > > 2. We already implement proxy notifications for push/PR events from external github to internal Jenkins master to trigger CI/CD jobs.
> > 
> > 
> > Got it. Does the proxy forward notifications both-ways (i.e. the PR will be notified with the build result)?
> > @Jakob-Naucke are you thinking in a similar setup for your s390x machine?
> 
> @wainersm no, I should think we'd just set up a slave instead which merely requires a tunnel from the master.

We tried this way. But it seems hard for a slave behind firewall to join a master in public network.  Do you have any tips?

**Comment by Jakob-Naucke** [comment]
> > @wainersm no, I should think we'd just set up a slave instead which merely requires a tunnel from the master.
> 
> We tried this way. But it seems hard for a slave behind firewall to join a master in public network. Do you have any tips?

@peterzcst not yet, I'm afraid, we haven't got it running yet :sweat_smile:

**Comment by peterzcst** [comment]
We made enough progress on TDX CI parts.  It's already done that CI jobs for both PR builds and master builds work for CCV0 branch, with CC+ Qemu + TDVF. With this, we are ready to integrate this pipeline in CCV0 CI, together with non-TEE CI.  Further, we would integrate more pipelines for TDX like CC+CLH+TD-SHIM, CC+DB+TD-SHIM in CCV0 CI. 

#2

**Comment by wainersm** [comment]
We already have a AMD SEV machine in the Jenkins' pool. @peterzcst and team has worked on the support to run the tests behind the firewall for Intel TDX. I assume that @Jakob-Naucke and team will be using the same approach. Therefore, can I close this issue? Or do you still think it is worth keeping it opened for further discussions?

**Comment by hbrueckner** [comment]
> We already have a AMD SEV machine in the Jenkins' pool. @peterzcst and team has worked on the support to run the tests behind the firewall for Intel TDX. I assume that @Jakob-Naucke and team will be using the same approach. Therefore, can I close this issue? Or do you still think it is worth keeping it opened for further discussions?

@wainersm I have requested an IBM z15 LPAR supporing Secure Execution (SE) behind a firewall. This request is still being processed. Once it is ready, I will get back to you regarding VPN access from the CI server side. For now, I think it is ok to close this PR.

Cc @stevenhorsman @BbolroC

**Comment by peterzcst** [comment]
> We already have a AMD SEV machine in the Jenkins' pool. @peterzcst and team has worked on the support to run the tests behind the firewall for Intel TDX. I assume that @Jakob-Naucke and team will be using the same approach. Therefore, can I close this issue? Or do you still think it is worth keeping it opened for further discussions?

hi let's still keep this open. We need add TDX related pipelines in Jenkin's pool.

**Comment by arronwy** [comment]
For TDX CI, we added a self hosted runner for `td-shim` project which can route TDX specific workflows to our TDX server.
We also enable a webhook for forked `kata-container` projects which can receive the PR requests and send the build results to github. 
Thanks @wainersm to create a new issue for create Intel TDX CI pipeline #32 to track the status.

**Comment by wainersm** [comment]
Closing as this was part of the 1st release.
